src/core/layout/qgslayoutitemmap.cpp

```diff
@@ -1090,8 +1104,44 @@ bool QgsLayoutItemMap::readPropertiesFromElement( const QDomElement &itemElem, c
   return true;
 }
 
+bool QgsLayoutItemMap::hasCustomFramePath() const
+{
+  if ( mAtlasClippingSettings->enabled() && mAtlasClippingSettings->clipFrame() )
+  {
+    QgsGeometry g( mLayout->reportContext().currentGeometry( crs() ) );
+    if ( !g.isEmpty() && g.type() == Qgis::GeometryType::Polygon )
+    {
+      return true;
+    }
+  }
+
+  return mItemClippingSettings->isActive();
+}
+
 QPainterPath QgsLayoutItemMap::framePath() const
 {
+  if ( mAtlasClippingSettings->enabled() && mAtlasClippingSettings->clipFrame() )
+  {
+    QgsGeometry g( mLayout->reportContext().currentGeometry( crs() ) );
+    if ( !g.isEmpty() && g.type() == Qgis::GeometryType::Polygon )
+    {
+      QPolygonF visibleExtent = calculateVisibleExtentPolygon( false );
+      QPolygonF rectPoly = QPolygonF( QRectF( 0, 0, rect().width(), rect().height() ) );
+
+      //workaround QT Bug #21329
+      visibleExtent.pop_back();
+      rectPoly.pop_back();
+
+      //create transform from map coordinates to layout coordinates
+      QTransform transform;
+      QTransform::quadToQuad( visibleExtent, rectPoly, transform );
+      g.transform( transform );
```

Comment:
could we use layoutToMapCoordsTransform().inverted() here to avoid the near-duplicate logic?

Reply:
@nyalldawson , I don't think so, because that function takes scenes as coordinate input, not the item positions (relative to the scene).